src/main/cocoa/CocoaDisplay.mm

```diff
@@ -0,0 +1,390 @@
+        namespace cocoa
+        {
+            CocoaDisplay::CocoaDisplay(): IDisplay()
+            {
+               bExit                   = false;
+            }
+
+            CocoaDisplay::~CocoaDisplay()
+            {
+            }
+
+            status_t CocoaDisplay::init(int argc, const char **argv)
+            {
+                [NSApplication sharedApplication];
+                [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
+                [NSApp activateIgnoringOtherApps:YES];
+                
+                return IDisplay::init(argc, argv);
+            }
+
+            status_t CocoaDisplay::main() {
+                while (!bExit) {
+                    timestamp_t ts = system::get_time_millis();
+
+                    // Do one main iteration
+                    status_t result = do_main_iteration(ts);
+                    
+                    if (result != STATUS_OK)
+                            return result;
+                }
+                
+                return STATUS_OK;
+            }
+
+            status_t CocoaDisplay::do_main_iteration(timestamp_t ts) {
+                // Here, any queued Cocoa events already handled via sendEvent.
+                // Use this to do your own rendering / app logic.
+                @autoreleasepool {
+                    // Process all pending Cocoa events
+                    NSEvent *event;
+                    while ((event = [NSApp nextEventMatchingMask:NSEventMaskAny
+                                                    untilDate:[NSDate distantPast]
+                                                        inMode:NSDefaultRunLoopMode
+                                                        dequeue:YES]))
+                    {
+                        [NSApp sendEvent:event];
+                        [NSApp updateWindows];
+
+                        //Dispatch to custom handler
+                        handle_event(event);
+                    }
+
+                    // Handle internal tasks
+                    status_t result = process_pending_tasks(ts);
+
+                    // Cocoa does not need explicit display flush like X11,
+                    // but you could call [NSApp windows] to force layout updates if needed
+
+                #ifdef USE_LIBFREETYPE
+                    sFontManager.gc();
+                #endif
+
+                    return result;
+                }
+                
+            }
+
+            void CocoaDisplay::handle_event(void *event)
+            {
+                NSEvent *nsevent = (__bridge NSEvent *)event;
+
+                if (!nsevent)
+                    return;
+
+                NSEventType type = [nsevent type];
+                
+                NSWindow *nsWindow = [nsevent window];
+                CocoaWindow *target = find_window(nsWindow);  // You must implement this
+
+                if (!target)
+                    return;
+                
+                event_t ue = {};
```

Comment:
We need to call `init_event` here to properly fill all fields of `ue` with default values.